docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,698 @@
+   ...
+
+   var fits bool
+   for {
+       // assume that all the msgs in the square use the non-interactive
+       // default rules and see if we can fit them in the smallest starting
+       // square size. We start the cursor (share index) at the beginning of
+       // the message shares (txShares+evdShares), because shares that do not
+       // follow the non-interactive defaults are simple to estimate.
+       fits, msgShares = shares.FitsInSquare(txShares+evdShares, squareSize, msgLens...)
+       switch {
+       // stop estimating if we know we can reach the max square size
+       case squareSize >= consts.MaxSquareSize:
+           return consts.MaxSquareSize, txShares + evdShares + msgShares
+       // return if we've found a square size that fits all of the txs
+       case fits:
+           return uint64(squareSize), txShares + evdShares + msgShares
+       // try the next largest square size if we can't fit all the txs
+       case !fits:
+           // increment the square size
+           squareSize = int(nextPowerOfTwo(squareSize + 1))
+       }
+   }
+}
+```
+
+#### Pruning excess transactions
+
+If there are too many transactions and messages in the square to fit in the max square size, then we have to remove them from the block. This can be complicated, as by default we want to prioritzie transactions that have higher fees, but removing a low fee transaction doesn't always result in using less shares.
+
+The simplest approach, and the one taken in the initial implementation, works by prematurely pruning the txs if we estimate that too many shares are being used. While this does work, and fulfills the constraints discussed earlier to create valid blocks, it is suboptimal. Ideally we would be able to identify the most optimal message and transactions to remove and then simply remove only those. As mentioned earlier, technically, a single byte difference could change the entire arrangement of the square. This makes arranging the square with complete confidence difficult not only because we have to follow all of the constraints, but also because of our frequent reliance on variable length length delimiters, and protobuf changing the amount of bytes used depending on the size of ints/uints.
+
+```go
+func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+   ...
+   // the totalSharesUsed can be larger that the max number of shares if we
+   // reach the max square size. In this case, we must prune the deprioritized
+   // txs (and their messages if they're pfd txs).
+   if totalSharesUsed > int(squareSize*squareSize) {
+       parsedTxs = prune(app.txConfig, parsedTxs, totalSharesUsed, int(squareSize))
+   }
+   ...
+```
+
+#### Malleation
+
+Due to the use of the `parsedTxs` data structure, we can now abstract the malleation process entirely, whereas before it was hard coded into the share proprietary proposal share splitting logic.
```

Comment:
> before it was hard coded into the share proprietary proposal share splitting logic

This sentence is hard to follow.

Reply:
good point: 69826cce25e69b231a2952692635acb41f537c06